Can get cached bitcoin balances and block info from CoreBitcoin (Issue #98)

willy-b opened the pull request:
To address #98 

Makes it so CoreBitcoin issues "balance" events even if balance is unchanged whenever it checks the balance, to support stateless/new consumers.

Commit:
CoreBitcoin issues "balance" events even if balance is unchanged, for stateless consumers

willy-b commented:
tl;dr I'm going back but storing the last value in DattCore

Full plan:
The way I "fixed" this was stupid: You have to wait the polling interval after opening/closing the settings pane. For performance reasons the polling interval should be kept long, but for UX reasons, the refresh interval should be short. Thus these should be totally decoupled.
Better solution: I'm going to keep the balance state at the DattCore level, and add a method to get the current balance. The maximum UI wait is then the polling interval, experienced ONLY if they opened the settings immediately before the first poll. *Almost always* one will get the last polled value instantly from DattCore.

Commit:
Cached bitcoin balances and block info

UI can be populated immediately from cache, most of the time (so does not wait on async calls unnecessarily)
- Cached balances are updated by polling, cached block info is updated whenever an async call is made.
- New events on DattCore ("bitcoin-block-info") and CoreBitcoin ("block-info") are emitted for fresh blocks

- Code changes:
  (+) DattCore/CoreBitcoin#getLastBalances (NEW) -> returns synchronously the current known state of the users bitcoin balances which are polled separately (this does NOT cause balances to be polled)

  (U) CoreBitcoin#asyncGetLatestBlockInfo -> now EMITS "block-info" event on CoreBitcoin if new block info is retrieved and caches the block info for synchronous retrieval by CoreBitcoin#getLastBlockInfo

  (+) DattCore/CoreBitcoin#getLastBlockInfo (NEW) -> returns synchronously a cached copy of last block info retrieved by CoreBitcoin#asyncGetLatestBlockInfo

  (U) DattCore#monitorCoreBitcoin -> now also EMITS "bitcoin-block-info" event on DattCore when new block info is loaded

  (+) BoxBitcoin#monitorDattCore -> NOW uses cached bitcoin balance and block info (if avail) to initialize UI while waiting for fresh async data
  -> uses DattCore#getLastBalances & DattCore#getLastBlockInfo to get the cached data

  (+) BoxBitcoin#handleBlockInfo (NEW) -> updates component state when a new block is received, right now just blockheightnum

- Tests:
  (+) DattCore/CoreBitcoin#getLastBalances (new test)

willy-b commented:
Almost ready for review. 
Just pending a couple test additions and an update:
  CoreBitcoin#asyncGetLatestBlockInfo: (need to update to verify that 'block-info' event is emitted)
  DattCore/CoreBitcoin#getLastBlockInfo: (new test needed)
  DattCore#monitorCoreBitcoin: (need to update to verify 'bitcoin-block-info' event is emitted)

Commit:
Test update: CoreBitcoin#asyncGetLatestBlockInfo verifies "block-info" event emitted

Tests:
 (U) CoreBitcoin#asyncGetLatestBlockInfo (updated to verify that block-info event is emitted)

Commit:
New test: CoreBitcoin#getLastBlockInfo

Tests that null is returned if no block info is cached, and that the block is cached after being retrieved by #asyncGetLatestBlockInfo

Commit:
New test DattCore#getLastBlockInfo

ryanxcharles commented on core/core-bitcoin.js:
```diff
@@ -138,8 +138,28 @@ CoreBitcoin.prototype.asyncPollBalance = function () {
+    let blockChanged = blockInfoFound && (!this.lastBlockInfo || !this.lastBlockInfo.hashbuf || this.lastBlockInfo.hashbuf.toString('hex') !== blockInfo.hashbuf.toString('hex'))
+
+    if (blockChanged) {
+      this.emit('block-info', blockInfo)
```
This seems like poor practice to me to emit an event inside a "get" method. The only thing the "get" function should do is get the value. Perhaps the proper place for this code to be is inside the blockchainAPI code, which could be set to monitor for the latest block info. And in that case, when an update is found, an event is emitted.

ryanxcharles commented on react/box-bitcoin.jsx:
```diff
@@ -49,7 +49,16 @@ let BoxBitcoin = React.createClass({
     let dattcore = this.props.dattcore
+    let initBalances = dattcore.getLastBalances()
+    if (initBalances) {
+      this.handleBitcoinBalance(initBalances)
```
The idea of the "monitor" function is to just have all the .on methods. There should probably be a different function for "emit initial events". Then the handle functions will handle them like normal. But then again, I'm not sure the UI should be emitting any initial events. The initial data should just be plugged into the props, no?

Commit:
Refactor CoreBitcoin block info updating and add tests

See RXC comment https://github.com/dattnetwork/datt/pull/101/files#r53862234

Code:
(+) CoreBitcoin#asyncUpdateBlockInfo: retrieve latest block info from BlockchainAPI#asyncGetLatestBlockInfo, cache in CoreBitcoin.lastBlockInfo, emit "block-info" as appropriate
(U) CoreBitcoin#asyncGetLatestBlockInfo: updated to call CoreBitcoin#asyncUpdateBlockInfo and then return up-to-date block info

Test:
(+) CoreBitcoin#asyncUpdateBlockInfo: test added
(U) DattCore#getLastBlockInfo: fix test cases and mock underlying BlockchainAPI call for speed

Commit:
BoxBitcoin#monitorDattCore - move balance/block-info initialization out

See RXC comment https://github.com/dattnetwork/datt/pull/101/files#r53863592

Code:
- BoxBitcoin#monitorDattCore: only "dattcore.on" methods here now
- BoxBitcoin#componentWillMount: try to set balance and block info from cache

N.B. We do not use the props since the block info and balance are updated in the state.
It is an anti-pattern to repeat data in props and state.

willy-b commented:
Thanks for the review. Addressed your comments, let me know if you need anything else

ryanxcharles commented:
Project has been rearchitected - PR no longer applies.
